### PR TITLE
Feature/747 custom metrics exporters

### PR DESCRIPTION
**Pull request by bastbu**
# PR for issue #747

## What is being addressed

With this PR we add support to raise custom metrics via the `System.Diagnostics.Metrics` API and export them to either Application Insights or on the Prometheus endpoint. We only add support for non-observable metrics (`Histogram<T>` and `Counter<T>`).

## How is this addressed

Since for both Application Insights and the prometheus-net SDK custom dimensions/tags need to be defined per metric, we introduce a registry of Custom Metrics that we track. We can only determine at runtime whether the metric raised has the correct custom dimensions/tags.

An example metric for the number of `JoinRequests` is added to the solution to illustrate how the registry can be used to raise metrics.

### CODE REVIEW

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs [hunk 1]
```diff
@@ -190,6 +195,7 @@ internal async Task InternalHandleDataAsync(RouteValueDictionary routeValues, We
                     break;
                 case LnsMessageType.JoinRequest:
                     LogReceivedMessage(this.logger, "jreq", json, null);
+                    JoinRequestCounter.Add(1, new[] { KeyValuePair.Create(MetricRegistry.GatewayIdTagName, (object)stationEui) });
```
Comment (danigian) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs:198]: We might want to track this metric later, after deduplication

Comment (bastbu) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs]: Addressed this in 98676615807f87182133e578e24f20c3ac20cd46

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs [hunk 2]
```diff
@@ -199,6 +204,8 @@ internal async Task InternalHandleDataAsync(RouteValueDictionary routeValues, We
                             break;
                         }
 
+                        JoinRequestCounter.Add(1, new[] { KeyValuePair.Create(MetricRegistry.GatewayIdTagName, (object)stationEui) });
```
Comment (Mandur) [inline LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs:207]: Should we track the join request after the mic was successfully checked? At this stage could still be from another network?

**Review by danigian — APPROVED** [review]

File: LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs (hunk 2 above)
Comment (bastbu) [reply LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/BasicsStation/Processors/LnsProtocolMessageProcessor.cs]: Addressed this in 07808a34fe97b4981c5d12ca4eeeb025a077d4af

**Pull request merged**